Add MultiStreamThroughput benchmark to object storage benchmarks

noahl opened the pull request:
Add the MultiStreamThroughput benchmark for object storage
services. This also requires splitting WriteObjects() in
scripts/object_storage_api_tests.py into two functions that work
together.

noahl commented:
@stfeng @yuyantingzero please take a look.

zbjornson commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -518,6 +593,148 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
 
 
+def MultiStreamThroughputBenchmark(storage_schema, host_to_connect=None):
+  """A benchmark test for single stream upload and download throughput.
```
s/single/multi/ :)

noahl replied:
Thanks :). Fixed.

zbjornson commented:
This is cool. It looks like you have _n_ workers read the same files concurrently, yes?

I'm hoping later that there will be another object_storage_service scenario that tests parallelized operations, a la https://github.com/rlmcpherson/s3gof3r for S3 and [composite objects/parallel uploads](https://cloud.google.com/storage/docs/composite-objects) for GCP -- that's actually what I'd think of with "multistream," vs. perhaps "concurrent" here. Not sure if the result would be much different though -- do you?

I don't think any GCS SDK or CLI provides parallel composite downloads (e.g. using the Range parameter to get parts of an object and reassemble client-side) yet to add what I'm describing.

zbjornson commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -518,6 +593,148 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
+
+  The object size is given by FLAGS.object_size_kb. The total number
+  of objects to transfer (summed across all streams) is given by
+  FLAGS.iterations. The benchmark starts at FLAGS.write_start_time, if
```
I don't see this flag defined at the top...

noahl replied:
Oops, I used that flag in an earlier version of this code before replacing it with `FLAGS.objects_per_stream`. Fixed.

zbjornson commented:
I see now, different files for each worker, which is yet another scenario than what I was thinking of... Pardon the noise.

stfeng commented:
Regarding parallel operations to the same object:

Gsutil has recently added a feature to download a large object in parallel,
Google: { gsutil Sliced Object Downloads} and its enabled by default with 4
"slices"

As to benchmarking support, pkb already supports CLI based parallelism: as
long as the default setting of a CLI tool uses parallelism, we will use it
in PKB too. (And take a look at the large object CLI test)

On Tuesday, February 9, 2016, Zach Bjornson <notifications@github.com>
wrote:

> This is cool. It looks like you have *n* workers read the same files
> concurrently, yes?
>
> I'm hoping later that there will be another object_storage_service
> scenario that tests parallelized operations, a la
> https://github.com/rlmcpherson/s3gof3r for S3 and composite
> objects/parallel uploads
> <https://cloud.google.com/storage/docs/composite-objects> for GCP --
> that's actually what I'd think of with "multistream," vs. perhaps
> "concurrent" here. Not sure if the result would be much different though --
> do you?
>
> I don't think any GCS SDK or CLI provides parallel composite downloads
> (e.g. using the Range parameter to get parts of an object and reassemble
> client-side) yet to add what I'm describing.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/pull/840#issuecomment-182170166>
> .
>

noahl commented:
The timing constants at the top of `object_storage_service_benchmark.py` are not yet ready for production. I'm still working out what the values should be. However, I'd appreciate a review of the rest of the code. Those are static variables that can be changed with no changes to anything else, so that should not make much difference.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -518,6 +593,148 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
+
+  logging.info('About to create processes for multistream.')
+
+  processes = [threading.Thread(target=MultiStreamThroughputWorker,
```
Please rename all processes to threads or something similar to that. I remember you used to use processes but changed it to be threads (which is correct and consistent with this code here) but let's also make sure we name the variable properly to reflect that change.

noahl replied:
Everything has been renamed (I think).

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -518,6 +593,148 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
+                                      host_to_connect,
+                                      FLAGS.object_size_kb,
+                                      FLAGS.objects_per_stream,
+                                      FLAGS.write_start_time,
```
For write and read start time, I'd like to suggest that these should not be plumbed into the thread workers. The workers should be made as dumb as possible and all they do is: write X number of objects or read X number of objects.

The intelligence of when to start write or read should be contained in the caller, that is, here.

Regarding the write_start_time and read_start_time design:

write-start-time is easy to implement: just start the writer threads when current time > write_start_time here (or start immediately if write_start_time is none).

Read_start_time would be slightly different: it's the responsibility of the caller here to wait for all writer threads to finish writing, and then it would launch the reader threads if current time > read_start_time.  this way, we will always get a very clean separation between the write phase and read phase among these threads on each VM.

Yes, above implies that we should split the thread worker into writer and reader so we have more flexibility in controlling them, and have better separation.

noahl replied:
After talking in person, I made the controlling PKB process ask for reads and writes separately. The write process sends the read process a list of already-written files by serializing it in `/tmp`, as discussed. This makes `object_storage_api_tests.py` much simpler.

stfeng commented on perfkitbenchmarker/linux_benchmarks/object_storage_service_benchmark.py:
```diff
@@ -192,6 +207,14 @@
+
+# The delay between when we start reading and when we start writing is
+# SECS_PER_UPLOAD_KB * # KB uploaded per thread + READ_DELAY_SECS
+SECS_PER_UPLOAD_KB = 0.01
```
If we use relative time to express read_start_time, then we won't need to get into this business of estimating how fast a provider can do uploads, which is quite fragile.

see more comments in the calculating write/read time logic

noahl replied:
Changed following our in-person conversation.

stfeng commented on perfkitbenchmarker/linux_benchmarks/object_storage_service_benchmark.py:
```diff
@@ -283,6 +306,58 @@ def _MakeSwiftCommandPrefix(auth_url, tenant_name, username, password):
+
+  write_start_time = time.time() + MULTISTREAM_BUFFER_TIME
+  kb_to_upload = FLAGS.object_size_kb * FLAGS.objects_per_stream
+  read_start_time = (write_start_time +
```
Let's see if it works better to express read_start_time as relative time, i.e., x seconds after all writes are complete.

When the test script on a VM gets invoked by us, it gets two time values: one is the write_start_time, which is expressed in absolute time, and then it does all the write tests using multiple streams, then it waits for all the writes to finish. After that, it checks if there is a "read_delay_time", if no, start reads immediately, if yes, wait for a read_delay_time and then start reads. This way, we would never have to guess how fast a provider can do upload to properly set the read_start_time as absolute timestamp.

To select a read_delay_time, it should be more than (perhaps 2x-3x more than) the maximum time it takes to launch the test script one by one on N VMs which is "relatively" less variable (because that would be mostly comprised of time to call RemoteCommand())

noahl replied:
Changed based on our in-person conversation.

stfeng commented on perfkitbenchmarker/linux_benchmarks/object_storage_service_benchmark.py:
```diff
@@ -55,7 +55,8 @@
 flags.DEFINE_enum('object_storage_scenario', 'all',
-                  ['all', 'cli', 'api_data', 'api_namespace'],
+                  ['all', 'cli', 'api_data', 'api_namespace',
+                   'api_multistream'],
```
Add comments below in the flag help text to clarify the difference between api_data and api_multistream. (yes, I know the help text as it stands is missing api_data)

noahl replied:
Done.

stfeng commented on perfkitbenchmarker/linux_benchmarks/object_storage_service_benchmark.py:
```diff
@@ -83,6 +84,20 @@
                      'Allow swiftclient to access Swift service without \n'
                      'having to verify the SSL certificate')
 
+flags.DEFINE_integer('objects_per_stream', 1000,
```
can we prefix these flags with "multistream" if they only apply to multistream.

noahl replied:
Yes, and I'll also prefix them with 'object_storage', which I forgot to do before!

noahl replied:
More thoughts: I don't want to prefix the object size flag with "multistream" because, even though it currently only applies to multistream, it clearly could apply to other scenarios, and it would be very natural to do so in the future. I've prefixed the other two with `multistream`.

stfeng replied:
SG. please change the helper text to reflect this. (currently it says applies to multi-streams only)

stfeng commented on perfkitbenchmarker/linux_benchmarks/object_storage_service_benchmark.py:
```diff
@@ -83,6 +84,20 @@
+                     'Size of objects to send and/or receive. Only applies to '
+                     'the api_multistream scenario.',
+                     lower_bound=1)
+flags.DEFINE_integer('num_streams', 1,
```
if this is for multi-stream, can we make the default to be more than 1 stream?

noahl replied:
Sure. The default is now 10.

stfeng commented on perfkitbenchmarker/linux_benchmarks/object_storage_service_benchmark.py:
```diff
@@ -192,6 +207,10 @@
 }
 AWS_S3_ENDPOINT_SUFFIX = '.amazonaws.com'
 
+MULTISTREAM_FIXED_DELAY = 5.0
```
Please add more comment to clarify why do we need these two delays, and what each one is for. My understanding is:

Multistream_fixed_delay is to accomondate multiple VM start time differences. it covers the most likely time it takes to launch remote commands to N VMs.

Multistream_delay_per_stream is to acommondate multiple stream start time differences on a single VM.

If my understanding is correct, two follow-up items:

1) please consider rename these variables to reflect this.
2) the 2 second per stream overhead is too high. This number is to accomondate  per thread start overhead so the last thread started can start write at the same time as the first thread that's started. I don't think it's going to take 2 seconds to start one thread. If I have to estimate, it will be far less than 100ms or so.

noahl replied:
I added comments and renamed the variables to reflect your comment above. I also changed the delay per stream to 0.1 sec for right now. We'll have to see how that works in experiments to know for sure the right value.

stfeng commented on perfkitbenchmarker/linux_benchmarks/object_storage_service_benchmark.py:
```diff
@@ -83,6 +84,20 @@
+                     'Number of objects to send and/or receive per stream. '
+                     'Only applies to the api_multistream scenario.',
+                     lower_bound=1)
+flags.DEFINE_integer('object_size_kb', 1,
```
Just to clarify, this PR does not include object size distribution support yet, right?

(in the future, when we add object size distribution support, how will that play together with the "api_multistream" scenario here?)

noahl replied:
Yes, this PR does not include object size distribution support.

However, I realize that I don't want to accidentally get a flag into a released version of PKB that I will have to rename when distributions lands I think the best choice would be to give it a forward-compatible name. I like `object_storage_object_size`. Right now, it will accept a value that is a single size, and 100% of the objects will be that size. When distribution support comes, the user will be able to supply either a distribution or a single size, which will be interpreted as a point distribution.

I think having the convenience of not specifying a full distribution when you want 100% of objects to be the same makes sense anyway as a convenience feature, so I don't think this adds unnecessary technical debt. How does that sound?

As to your question about how that will play together with the "api_multistream" scenario, I am planning for each stream to choose objects from the distribution independently. I will document that when I add distribution support.

stfeng replied:
Currently I see only the SingleStream test that will likely use this object_size flag. Other test scenarios (namespace, one byte RW, or cli) all have their own object sizes. If we call it object_storage_object_size, people would think we'd use this object size for all scenarios, which is not true. Consider a different name...

Alternatively, since you are going to make a second PR to support object size distribution very soon, I am leaning towards not introduce this new flag that specifies a single value object size. Once you add support for object size distribution, then it's very easy to specify a single size via the distribution flag, right? (something like "200MiB, 100%")

stfeng replied:
We discussed in person and the conclusion is we will use object_sizes as the flag name, for now it support one single size, and later it will be expanded to support a distribution.

stfeng commented on perfkitbenchmarker/linux_benchmarks/object_storage_service_benchmark.py:
```diff
@@ -438,6 +490,37 @@ def BuildBenchmarkScriptCommand(args):
+      write_start_time = (time.time() +
+                          MULTISTREAM_FIXED_DELAY +
+                          MULTISTREAM_DELAY_PER_STREAM * FLAGS.num_streams)
+
```
Add some info logs here to indicate the progress of this test:

For example, add progress information at each step start/stop time.
log.info("starting write multi-stream test")

log.info("write mult-stream test completed successfully, starting read multi-stream test...)

etc.

noahl replied:
Added.

stfeng commented on perfkitbenchmarker/linux_benchmarks/object_storage_service_benchmark.py:
```diff
@@ -865,7 +948,7 @@ def Prepare(self, vm):
     vm.bucket_name = 'pkb%s' % FLAGS.run_uri
 
-    vm.RemoteCommand('%s mb gs://%s' % (vm.gsutil_path, vm.bucket_name))
+    vm.RemoteCommand('%s mb -l ASIA gs://%s' % (vm.gsutil_path, vm.bucket_name))
```
what's this line for? Un-intended change for this PR? ;)

noahl replied:
Yes, that was a precursor to #845 . It's gone from this PR :).

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -57,7 +61,8 @@
 flags.DEFINE_enum(
     'scenario', 'OneByteRW', ['OneByteRW', 'ListConsistency',
-                              'SingleStreamThroughput', 'CleanupBucket'],
+                              'SingleStreamThroughput', 'CleanupBucket',
```
add help text to the flag documentation.

noahl replied:
Added.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -67,6 +72,21 @@
                      'particular test scenario. Currently only applicable to '
                      'the ListConsistency scenario, ignored in others.')
 
+flags.DEFINE_integer('objects_per_stream', 1000, 'The number of objects to '
```
Same comments to these flags as I made in the pkb "driver" script. 

I think we should be careful about module layering here. We need to think (clearly) about how we plan to support object size distribution and number of streams. Even though we don't have the object size distribution support in this PR, it's going to be helpful to the (very) near future when you add such support.

My thought (open for discussions):
1. We should add object size distribution support to this file here, it will take a distribution config file (# of streams, objects per stream, and size distribution), or a set of flags specifying the distribution (per design doc), and work off on that. 
2. The "driver" script will either pass through user's test configure file straightly to this file (after some sanity check), or create a config file by itself for the api_multi_stream scenario based on the 3 flags user passed-in for that scenario.

If that's the direction we want to head for, then we should change the flags here (especially the name) to reflect the end state when they are used to describe a distribution. We don't have to implement the full distribution here in this PR, (e.g., we can assume 100% object will be of the same size). But we do want to get the flag names (and meaning) correct so we don't need to change the flags next time when we add distribution support.

noahl replied:
In the big commit I pushed today, I followed your suggestion and added the outline of object size distribution support. Here is how it works at a high level:

- the user supplies a distribution to the benchmark using `--object_storage_object_sizes=...`. The distribution can either be a single size, like "1KB", or a dictionary of sizes and percentages, like "{10KB: 10%, 100KB: 90%}".
- the distribution is passed to `_DistributionToBackendFormat`, which converts it to a slightly simpler format: all object sizes are integers, representing the size in bytes, and the frequencies are floating-point numbers.

It adds complexity to have two different representations of distributions, but I think it's worth it because it lets the user specify distributions using units, but lets us keep pint out of the worker (I especially wanted this because we have to customize pint's unit system slightly, and I don't want to keep the customizations in sync in two places). This also lets us pass distributions to the backend as a JSON data structure, which means we can use the json library that comes with Python to serialize and deserialize, so we don't add any dependencies in the worker.
- The worker script accepts a distribution in the "backend format" and turns that into a Python iterator object, which can be used to draw from the distribution with `next()`.

This is the only part where full distribution support is not complete: the iterator object requires that the distribution has only one size in it. I think this design is going to make it very easy to add full distribution support in a future PR.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -117,6 +137,11 @@
 
+BYTES_PER_KILOBYTE = 1024
+
+# If we would sleep less than 0.1 seconds, don't bother.
```
why introduce this variable? If there is no clear "badness" of sleeping for less than 100ms, then we shouldn't need to have this.

noahl replied:
Okay, I'll just remove it.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -244,6 +269,19 @@ def CleanupBucket(storage_schema):
+
+  payload_string = payload_bytes.decode('ascii')
+
+  return payload_bytes, payload_string
```
why do we need to return both payload_bytes and payload_string? It takes 2x the memory. We should only need one for subsequent usage.

noahl replied:
I did it because I was keeping the interface exactly the same as before. I strongly suspect the benchmark currently uses 2x the memory necessary. Looking at the code, the reason is that the Azure backend uses the bytes variable, and the boto backend uses the string variable.

I'm preparing a separate PR in which I upgrade our S3 support to boto3 (it's preparation for supporting multiple storage classes). As part of that PR, I plan to move different backends into different classes to use Python's polymorphism instead of `if` statements and `_useBotoApi()` (there will be 3 backends then, and I think that will be much more clean). I think that is the most natural place to fix this issue.

noahl replied:
I will plan to submit that PR first and then rebase this on top of that, so the problem will go away before I submit this.

stfeng replied:
I am fine with returning both payload_bytes and payload_string as is in this PR. (it's the same behavior today).

Good to know that you are working on a separate PR that will refactor this to 3 "backends", but that is probably not required for this PR to work. For now, I'd stay focused on getting this one and the object size distribution done first.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -518,6 +592,178 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
+def WriteWorker(storage_schema, host_to_connect,
+                object_size_kb, num_objects,
+                start_time, result_queue, worker_num):
+  payload_bytes, payload_string = GenerateWritePayload(
```
we shouldn't have each worker generate this payload, all workers should share this read-only payload buffer.

noahl replied:
Fixed! Ouch, that was a big bug.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -266,17 +304,44 @@ def WriteObjects(storage_schema, bucket, object_prefix, count,
+    count: The total number of objects that need to be written.
+    payload_bytes: A bytearray to transfer.
+    payload_string: A string version of payload_bytes.
+    objects_written: A list of names of objects that have been successfully
```
Add comment for the "size" parameter.

noahl replied:
Added.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -266,17 +304,44 @@ def WriteObjects(storage_schema, bucket, object_prefix, count,
+                           payload_bytes, payload_string, size,
+                           objects_written, latency_results=None,
+                           bandwidth_results=None,
+                           object_start_times=None,
```
Minor: consider rename "object_start_times" to "operation_start_times"?

noahl replied:
How about just `start_times`?

stfeng replied:
SG.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -326,8 +395,10 @@ def ReadObjects(storage_schema, bucket, objects_to_read, latency_results=None,
     latency_results: An optional list to receive latency results.
     bandwidth_results: An optional list to receive bandwidth results.
     object_size: Size of the object that will be read, used to calculate bw.
+    object_start_times: An optional list to receive start time results.
```
same comment as above...

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -518,6 +592,178 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
                   host_to_connect=host_to_connect)
 
 
+def RunMultiThreadedWorkersWriteResults(worker, worker_args,
```
minor: how about just "RunThreadedWorkers"? (threaded implies multi-threaded, "write results" is an implementation detail and kind of implied too: we will get results out of this function)

noahl replied:
Changed. Your name is much better.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -518,6 +592,178 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
+
+  Args:
+    worker: either WriteWorker or ReadWorker. The worker function to call.
+    worker_args: a tuple. The arguments to pass to the worker function. The
```
Is this a list or a tuple?

noahl replied:
I do pass in a tuple, so the documentation is correct here. According to the Python threading docs, a tuple is the correct type.

stfeng replied:
SG.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -518,6 +592,178 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
+
+  result_queue = Queue.Queue()
+
+  logging.info('Creating threads')
```
log how many threads we are about to create.

noahl replied:
Done.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -518,6 +592,178 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
+    thread.start()
+  logging.info('Threads started. Joining threads.')
+  for thread in threads:
+    thread.join()
```
what's the default timeout for the join()? Any thoughts on how we prevent hang?

noahl replied:
The default is no timeout. I think that ideally each thread would send updates on its progress to the master thread, and we would kill a thread that wasn't making progress.

A less good but easier to implement solution is to poll threads to see if they are finished, and once some percentage are done, give the rest a fixed time interval to finish before we kill them. This would protect against the situation where most threads are okay but one or two are stuck.

An even less good alternative, which I think we should take for v1, is to let the user kill the benchmark if it takes too long.

What do you think of those ideas?

stfeng replied:
As discussed in person, we will let users kill the benchmark if it's taking too long but we will print out progress information in the controller thread so users know we are making (or not making) progress.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -518,6 +592,178 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
+    thread.join()
+  logging.info('All threads complete.')
+
+  try:
```
Note to myself: continue from here...

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -518,6 +592,178 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
+        json.dump({'kind': res['kind'],
+                   'start_time': start_time,
+                   'latency': latency,
+                   'size': FLAGS.object_size_kb,
```
We should have the size be part of the result too. This will enable the upcoming support of object size distribution.

noahl replied:
Done.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -518,6 +592,178 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
+      num_results += 1
+      stream_num = res['stream_num']
+      for start_time, latency in zip(res['start_times'], res['latencies']):
+        json.dump({'kind': res['kind'],
```
consider rename 'kind' to "operation".

noahl replied:
Done, here and in the read and write workers.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -518,6 +592,178 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
+    pass
+
+  if num_results != FLAGS.num_streams:
+    raise Exception('Couldn\'t get results from all streams!')
```
instead of raising the exception (e.g., 1 out of 100 stream has failed), can we return whatever we have collected successfully so far to the caller? with these two numbers: num_results and FLAGS.num_streams.

We should probably let the caller decide what to do in the case of partial failure.  (e.g., caller may decide to accept the results if there is only a very small fraction of streams that have failed)

noahl replied:
Yes, I'll just return the results. I don't think we have to pass back num_results because the results already include stream numbers, so the caller can figure out how many streams are there.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -518,6 +592,178 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
+  Args:
+    when: float. The time to sleep to, as a POSIX timestamp.
+
+    too_late: None or Exception. If when is before now + MIN_SLEEP,
```
My recommendation for the too_late case is simply log and ignore it, return immediately. This makes the benchmarking code more robust. There is no need to throw exception or return an error here - having all streams to start at about the same time is a nice to have, but not absolutely a must-to-have. Our post-run analysis script should be able to tolerate some degree of different start times.

noahl replied:
Done. That makes it much simpler.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -518,6 +592,178 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
+  objects_filename = os.path.join(TMP_DIR,
+                                  OBJECTS_WRITTEN_FILE_PREFIX + str(worker_num))
+
+  # Error handling deliberately omitted - abort on any exception.
```
can you catch and log the error, and then raise?

noahl replied:
I ended up removing this code.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -518,6 +592,178 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
+    SleepUntilTime(start_time, too_late=Exception)
+
+  WriteObjectsFromBuffer(storage_schema, FLAGS.bucket,
+                         ('pkb_write_worker_%s' % worker_num),
```
to differentiate object prefixes across different invocations, we also need to introduce current timestamp to the prefix

e.g., pkb_write_worker_1_{current_time}

This way, if someone invokes write multiple times (for any legitimate reason), they don't need to worry about those writes clobbing each other.

noahl replied:
I believe our latest plan is to support write-once, read-many-times by saving state from only the last write operation. Therefore, this comment is out of date.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -518,6 +592,178 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
+                         object_start_times=start_times,
+                         latency_results=latencies,
+                         host_to_connect=host_to_connect)
+
```
Add a comment here to detail:
1) why do we need to write the object names to a file.
2) what is the format of the file, and how the files are named.

noahl replied:
The comment has been added, although it's elsewhere in the file.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -518,6 +592,178 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
+  if os.path.exists(objects_filename):
+    os.remove(objects_filename)
+  with open(objects_filename, 'w') as out:
+    json.dump(objects_written, out)
```
In order to support object size distribution in the (near) future, please consider writing 2-tuples here to the file: {object_name, object_size}, although with this PR, the object_size will all be the same.

stfeng replied:
Is this one done?

noahl replied:
Yes, I write (object_name, object_size) to the state file.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -518,6 +592,178 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
+  with open(objects_filename, 'w') as out:
+    json.dump(objects_written, out)
+
+  result_queue.put({'kind': 'upload',
```
Consider putting the "object_size" as part of the result too.

noahl replied:
Done.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -518,6 +592,178 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
+  objects_filename = os.path.join(TMP_DIR,
+                                  OBJECTS_WRITTEN_FILE_PREFIX + str(worker_num))
+
+  # Error handling deliberately omitted - abort on any exception.
```
same comment as above, log the exception and then raise.

noahl replied:
This is obsolete given my refactor.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -518,6 +592,178 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
+                                  OBJECTS_WRITTEN_FILE_PREFIX + str(worker_num))
+
+  # Error handling deliberately omitted - abort on any exception.
+  with open(objects_filename, 'r') as inp:
```
variable naming: "inp", consider use full name.

stfeng replied:
is this one done or obsolete now?

noahl replied:
This is obsolete. I refactored the code to remove the line in question.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -518,6 +592,178 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
+
+
+def ReadWorker(storage_schema, host_to_connect,
+               object_size_kb, start_time,
```
for the ReadWorker, we don't need to pass in object_size_kb as parameter, it should be able to pick-up the object sizes for all objects from the temporary file, which was written by the writer.

stfeng replied:
Please refer to my latest comment regarding this.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -518,6 +592,178 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
+  ReadObjects(storage_schema, FLAGS.bucket, objects_written,
+              object_start_times=start_times,
+              latency_results=latencies,
+              object_size=object_size_kb * BYTES_PER_KILOBYTE,
```
later on when we support a distribution of object sizes, we will need to change the ReadObjects() function to take in an array of object_sizes.

noahl replied:
I don't use the bandwidth information that ReadObjects() provides anyway, so I refactor this to make ReadObjects() a wrapper method around an inner function that does not need size information. The multistream throughput code uses the inner function directly. This makes the future interface to distributions much simpler, and I also find this approach cleaner.

stfeng replied:
How do you plan to compute aggregated throughput in the case of multistream with single object size and multistream with object size distributions? Does it still involve computing throughput values for individual objects? If so, then you might still want to call this function with sizes information here.

noahl replied:
I will need to compute the per-object throughput, but I am planning to do that in an analysis phase after the benchmarking is complete. I will need an analysis phase no matter what, and since I need to have one, it's simple to have all of my analysis there. All I need from the worker is to collect the raw data, which is `start_time` and `latency` for each operation.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -518,6 +592,178 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
+              object_size=object_size_kb * BYTES_PER_KILOBYTE,
+              host_to_connect=host_to_connect)
+
+  result_queue.put({'kind': 'download',
```
add object sizes array along with the latencies to this result set, so caller can compute throughput.

noahl replied:
Done.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -518,6 +592,178 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
+OBJECTS_WRITTEN_FILE_PREFIX = 'pkb-objects-written-'
+
+
+def WriteWorker(storage_schema, host_to_connect,
```
For both read and writer worker code below, consider add a few more logging lines to indicate what they are doing or about to do: "going to sleep until xyz", "done sleeping, about to do work", etc, etc.

noahl replied:
Added.

stfeng commented on tests/object_storage_service_benchmark_test.py:
```diff
@@ -0,0 +1,78 @@
+  def setUp(self):
+    self.raw_result = """{"latency": 0.1, "kind": "download", "stream_num": 1, "start_time": 10.0, "size": 1}"""  # noqa: line too long
+
+  def testParseSingleResult(self):
```
How about add a test case for more than one result in the result array?

noahl replied:
Done.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -518,6 +592,178 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
+                                  OBJECTS_WRITTEN_FILE_PREFIX + str(worker_num))
+
+  # Error handling deliberately omitted - abort on any exception.
+  if os.path.exists(objects_filename):
```
A couple of additional questions regarding the object name files under /tmp:
1. Who is going to cleanup these files? Do we need to add logic in CleanupBucket() function?

2.I think we should remove all existing object state files under the /tmp directory, regardless of the worker number suffix. The use case we support is write once with X streams, and read it many times with Y1, Y2, ... Yn streams. If you try to do write tests more than once, the last write test will be retained and used for subsequent read test. 

This is simple and deterministic.

stfeng replied:
Is this one considered? The most interesting question is cleanup...

noahl replied:
Sorry I didn't reply to this earlier. I added cleanup to the `Cleanup()` method in `object_storage_service_benchmark.py`.

The writer also cleans up an old state file if it finds one before it writes a new one. I think that addresses both of your points.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -518,6 +592,178 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
+               object_size_kb, start_time,
+               result_queue, worker_num):
+
+  objects_filename = os.path.join(TMP_DIR,
```
As we discussed in person, we should support the use case where we need to write once and read multiple times with different stream counts. For that to work, we need to read all object name files under the /tmp dir (read the object name, size pairs into an array) and then sub-divide that array into segments and pass those segments into each read worker. That will be done by the caller of this function, this function will just get an array of <object_name, object_size> pairs and work on that.

noahl replied:
I ended up changing the logic so that the controlling thread writes one log file for all workers, rather than each individual worker saving its own file. Since the controlling process gets all the result data anyway to pass back to PKB, it's not really any more information flow, and it makes the log file logic much simpler. It also makes the worker thread functions small, which I like.

stfeng replied:
SG.

stfeng commented:
Noah, I made a couple of additional comments for the object name files that are stored under the /tmp directory (based on our in person chat). Please take a look.

Let me know when you are done incorporating my first round of CR feedback. I will make another pass. Thanks.

hildrum commented:
Should we wait for this to pull it into 1.3?

stfeng commented:
No need to wait on this one.

On Friday, February 26, 2016, Kris Hildrum <notifications@github.com> wrote:

> Should we wait for this to pull it into 1.3?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/pull/840#issuecomment-189386763>
> .
>

noahl commented:
@stfeng I've updated this in response to your review. Please take a look.

stfeng commented:
I am going to take a pass on the new code, but before that, here are a few un-answered comments from round 1 of CR. Probably you have already addressed them, if so please update these in the tool here:
=======
<copy-pasted from above un-answered comments>

1. In order to support object size distribution in the (near) future, please consider writing 2-tuples here to the file: {object_name, object_size}, although with this PR, the object_size will all be the same.

2. variable naming: "inp", consider use full name.

3. A couple of additional questions regarding the object name files under /tmp:
- Who is going to cleanup these files? Do we need to add cleanup logic in the cleanup function where we currently cleanup objects in the bucket?

noahl commented:
@stfeng I'm sorry I didn't reply to those comments before! I've gone back and added replies. Here they are again, all in one place:

1. I now write 2-tuples to the state file as you suggested.
2. I refactored the code so the `inp` variable no longer exists.
4. I added cleanup logic in `object_storage_service_benchmark.Cleanup()`.

voellm commented:
@noahl / @stfeng 
Looks like there are > 100 comments on this pull request.  Should we take this offline and have a meeting to review the design and issues?  Happy to include @zbjornson as well in the meeting if he would like to be there.

stfeng commented:
@voellm, we've already had a couple of f2f discussion meetings. Noah's latest commit here reflects the outcome of these discussions. I expect the focus of CR from this point on would be mostly "fit-and-finish" type of comments. If there were to be major comments, we will sure to talk and discuss again just like we have been doing.

stfeng commented on perfkitbenchmarker/linux_benchmarks/object_storage_service_benchmark.py:
```diff
@@ -86,6 +94,19 @@
+                     lower_bound=1)
+flag_util.DEFINE_yaml('object_storage_object_sizes', '1KB',
+                      'Size of objects to send and/or receive. Only applies to '
+                      'the api_multistream scenario. Example: {1KB: 100%}')
```
The example here is for how to specify a distribution. Please clarify that you can also specify a single size and give an example here

noahl replied:
Done.

stfeng commented on perfkitbenchmarker/linux_benchmarks/object_storage_service_benchmark.py:
```diff
@@ -202,6 +223,18 @@
+MULTISTREAM_DELAY_PER_STREAM = 0.1
+
+# The multistream write benchmark writes a file in the VM's temporary
+# storage with the objects it has written, which is used by the
```
s/"temporary storage"/"/tmp"/

noahl replied:
Done.

stfeng commented on perfkitbenchmarker/linux_benchmarks/object_storage_service_benchmark.py:
```diff
@@ -293,9 +326,145 @@ def _MakeSwiftCommandPrefix(auth_url, tenant_name, username, password):
+
+  for record in records:
+    metric = 'Multi-stream %s KB %s' % (record['size'], record['operation'])
+    md = metadata.copy()
```
why do you make a copy of the metadata? we should store the metadata in the original metadata if it's passed in.

noahl replied:
The original metadata contains fields that are common to all of the samples, but each sample will have a different stream_num and object_size. So I make a copy of the common metadata dictionary and update it with the unique fields for each sample.

stfeng replied:
Ah... we can't store each and every raw data point as "sample" here, that would be way too many. (e.g., 100 stream with 1000 object per stream would mean 100,000 samples *per run*)

In our current code, we have always performed aggregation on the raw results and then report aggregated numbers as samples. (e.g., average, count, mean, p5, p50, p90, etc)

Please take a look at _AppendPercentilesToResults() see how the rest of the code base does that, it's pretty simple to do the same here. 

As to metadata, for each sample point from the same run, they should have exactly the same metadata for this test case: how many streams were used in the test, and what is the object_storage_object_sizes parameter.

stfeng commented on perfkitbenchmarker/linux_benchmarks/object_storage_service_benchmark.py:
```diff
@@ -293,9 +326,145 @@ def _MakeSwiftCommandPrefix(auth_url, tenant_name, username, password):
+  results = []
+
+  for record in records:
+    metric = 'Multi-stream %s KB %s' % (record['size'], record['operation'])
```
Please consider include the keyword "latency" as part of the metric name here.

And, I don't think we should include size in the metrics name, it's should be part of the metadata (which shows up as labels in the result), this is because we allow user to specify the size of objects when they run this benchmark, so including the size in the metric name will lead to a proliferation of metric names. Each metric name should represent one scenario (or sub-scenario) that's being measured. In this case, we should have "Multi-stream upload latency", "Multi-stream upload throughput", etc. with number of streams, object sizes as metadata.

One question: we have latency metrics here, will you add throughput metrics as the next PR that does some post-run analysis of aggregated throughputs?

noahl replied:
Metric names are changed to "Multi-stream {upload|download} latency".

Yes, I can add throughput metrics in the next PR, but I'd like to think more about the design. I have tried very deliberately to store the raw unprocessed data as samples, so I will have as much flexibility as possible in the analysis step. I think it makes sense to also publish processed values as samples, but it's worth talking about.

noahl replied:
I also made sure that num_streams and objects_per_stream were part of the metadata. They weren't previously.

stfeng replied:
See my response above. The "analysis" step should be done after raw data collection (so you have the max flexibility), and before publishing the samples. The samples should be the results/output of your analysis step, not as an input to it.

stfeng commented on perfkitbenchmarker/linux_benchmarks/object_storage_service_benchmark.py:
```diff
@@ -293,9 +326,145 @@ def _MakeSwiftCommandPrefix(auth_url, tenant_name, username, password):
+  return results
+
+
+def _StringToBytes(string):
```
move this function and the next one to a utility file, to be shared and to benefit all other benchmarks. This is general enough to be out of this file.

noahl replied:
Done.

stfeng commented on perfkitbenchmarker/linux_benchmarks/object_storage_service_benchmark.py:
```diff
@@ -293,9 +326,145 @@ def _MakeSwiftCommandPrefix(auth_url, tenant_name, username, password):
+    # Catching all exceptions is ugly, but we don't know what sort of
+    # exception pint might throw, and we want to turn any of them into
+    # ValueError.
+    raise ValueError("Couldn't parse object size %s" % string)
```
remove "object" in the message.

noahl replied:
Done.

stfeng commented on perfkitbenchmarker/linux_benchmarks/object_storage_service_benchmark.py:
```diff
@@ -293,9 +326,145 @@ def _MakeSwiftCommandPrefix(auth_url, tenant_name, username, password):
+  try:
+    bytes = quantity.m_as(perfkitbenchmarker.UNIT_REGISTRY.bytes)
+  except pint.DimensionalityError:
+    raise ValueError("Quantity %s is not an object size" % string)
```
remove "object" from the message.

noahl replied:
Done.

stfeng commented on perfkitbenchmarker/linux_benchmarks/object_storage_service_benchmark.py:
```diff
@@ -293,9 +326,145 @@ def _MakeSwiftCommandPrefix(auth_url, tenant_name, username, password):
+    raise ValueError("Size %s has a non-integer number (%s) of bytes!" %
+                     (string, bytes))
+
+  if bytes < 0:
```
this does not handle integer overflow. anything more than 2GB will come to here. we need to support large objects. At least we should be able to support to the max of integer which is 4GB.

noahl replied:
According to http://stackoverflow.com/questions/4581842/python-integer-ranges , and verified by my own experiments, Python ints are automatically promoted to longs when they overflow, rather than wrapping around. Also, the `int` conversion function will return a long rather than raising an exception if its argument (I tested strings and floats) doesn't fit in an int. Therefore, I think this code will work correctly, except that `StringToBytes` may sometimes return a long when it normally returns an int. This could only be an issue when the user specifies an object with more than 2^63 bytes, which is 8 exabytes. I don't think we need to worry about that case.

stfeng replied:
SG. Thanks for checking on this.

stfeng commented on perfkitbenchmarker/linux_benchmarks/object_storage_service_benchmark.py:
```diff
@@ -293,9 +326,145 @@ def _MakeSwiftCommandPrefix(auth_url, tenant_name, username, password):
+    string: the object size, as a string with a quantity and a unit.
+
+  Returns:
+    an integer. The number of bytes in the size.
```
We need to talk about this.... but restricting possible object size to an integer is not ideal. Perhaps we can fix it in the next PR but make sure to file a git issue and assign to yourself if you decide to punt this.

noahl replied:
According to my experiements, Python ints should hold sizes up to 2^63 bytes = 8 exabytes, so I think this is fine. I did originally support longs, and I only removed that support when I ran into an issue with printing them out and reading them back in, because of the "L" on the end of their printed representation. I can fix that, but I think that ints work for large enough sizes that it shouldn't be an urgent issue.

stfeng replied:
SG.

stfeng commented on perfkitbenchmarker/linux_benchmarks/object_storage_service_benchmark.py:
```diff
@@ -448,6 +617,57 @@ def BuildBenchmarkScriptCommand(args):
+          MULTISTREAM_DELAY_PER_VM +
+          MULTISTREAM_DELAY_PER_STREAM *
+          FLAGS.object_storage_multistream_num_streams)
+
```
log the write_start_time here.

noahl replied:
Done.

stfeng commented on perfkitbenchmarker/linux_benchmarks/object_storage_service_benchmark.py:
```diff
@@ -448,6 +617,57 @@ def BuildBenchmarkScriptCommand(args):
+          MULTISTREAM_DELAY_PER_VM +
+          MULTISTREAM_DELAY_PER_STREAM *
+          FLAGS.object_storage_multistream_num_streams)
+
```
log read start time here.

noahl replied:
Done.

stfeng commented on perfkitbenchmarker/linux_benchmarks/object_storage_service_benchmark.py:
```diff
@@ -448,6 +617,57 @@ def BuildBenchmarkScriptCommand(args):
+          '--start_time=%s' % write_start_time,
+          '--objects_written_file=%s' % objects_written_file,
+          '--scenario=MultiStreamWrite'])
+      write_out, _ = vm.RemoteCommand(multi_stream_write_cmd, should_log=True)
```
Consider using RobustRemoteCommand here and below.

noahl replied:
Done.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -57,16 +63,37 @@
+flags.DEFINE_integer('objects_per_stream', 1000, 'The number of objects to '
+                     'read and write per stream in the MultiStreamThroughput '
+                     'scenario.')
+flags.DEFINE_string('object_sizes', 1, 'The size of the objects to use, as a '
```
Give examples here for the object_sizes flag, just like you did in the high level benchmarking script.

stfeng replied:
Also, hopefully the examples will show that we accept "bytes" only here, not KB, MB, etc.

noahl replied:
Done. The examples correspond to the two examples in `object_storage_service_benchmark.py`. I think they should show that we only accept bytes and what the backend distribution format is.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -57,16 +63,37 @@
+flags.DEFINE_integer('num_streams', 1, 'The number of streams to use. Only '
+                     'applies to the MultiStreamThroughput scenario.',
+                     lower_bound=1)
+flags.DEFINE_string('objects_written_file', None, 'The path where the '
```
In the help text, please mention that if there is an existing file there and write scenario is used, the existing file will be removed.

noahl replied:
Done.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -141,6 +174,40 @@ def _useBotoApi(storage_schema):
+    self.size = size
+
+  def __iter__(self):
+    return self  # this is required by the Python iterator protocol
```
move the trailing comment right above the function def.

noahl replied:
Moved.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -266,57 +346,74 @@ def WriteObjects(storage_schema, bucket, object_prefix, count,
-
-      latency = time.time() - start_time
+      _, latency = WriteObjectFromBuffer(
+          storage_schema, bucket, object_prefix,
```
(Big) BUG: object_prefix should be "object_name"

noahl replied:
Ouch! Fixed.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -266,57 +346,74 @@ def WriteObjects(storage_schema, bucket, object_prefix, count,
-                        object_name)
+
+def WriteObjectFromBuffer(storage_schema, bucket, object,
+                          payload_bytes, payload_string, size,
```
consider renaming the size to be more specific: "object_size", "size_to_write", etc which is <= the buffer's size (payload_bytes)

noahl replied:
Renamed "object" -> "object_name", "bucket" -> "bucket_name" and "size" -> "size_to_write". I think those were the ambiguous ones.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -266,57 +346,74 @@ def WriteObjects(storage_schema, bucket, object_prefix, count,
-      logging.exception('Caught exception while writing object %s.',
-                        object_name)
+
+def WriteObjectFromBuffer(storage_schema, bucket, object,
```
consider renaming "object' to "object_name"

noahl replied:
Done.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -266,57 +346,74 @@ def WriteObjects(storage_schema, bucket, object_prefix, count,
+    if host_to_connect is not None:
+      object_uri.connect(host=host_to_connect)
+
+    object_uri.set_contents_from_string(payload_string[:size])
```
Python does "slice-by-copy" so payload_string[:size] would generate a new array object and has a copy of the original array, which is very inefficient here. I think your design calls for sharing the read-only global buffer.

We need a "view" into the string or byte array, perhaps the memoryview class fits well here, please take a look at that.

noahl replied:
Fixed. memoryview is the solution. Thank you very much!

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -352,6 +445,38 @@ def ReadObjects(storage_schema, bucket, objects_to_read, latency_results=None,
       logging.exception('Failed to read object %s', object_name)
 
 
+def ReadObject(storage_schema, bucket, object, host_to_connect=None):
```
consider renaming "object" to "object_name"

noahl replied:
Done. I also renamed "bucket" to "bucket_name", because it matches and is more descriptive.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -518,6 +643,240 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
+    for result in results:
+      for name, size in zip(result['object_names'], result['sizes']):
+        object_records.append([name, size])
+    if os.path.exists(FLAGS.objects_written_file):
```
[Not a major issue] 
please consider wrapping these file IO lines with try-except, so if we fail to write to the object names file for any reason, we still get to return the results from our successful benchmarking runs to our caller. but we do need to log a big warning to indicate that file writing has failed, and don't issue read test unless they have an alternative/good object name file.

noahl replied:
Done.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -518,6 +643,240 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
+
+
+def MultiStreamWrites(storage_schema, host_to_connect=None):
+  """Run multi-stream write threads."""
```
Add more documentation to this function. (parameters, return type, etc)

noahl replied:
Added a big docstring.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -518,6 +643,240 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
+
+
+def MultiStreamReads(storage_schema, host_to_connect=None):
+  """Run multi-stream read threads."""
```
function documentation here.

noahl replied:
Added.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -518,6 +643,240 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
+                     'names to read from. Use '
+                     '--objects_written_file=<filename>.')
+  with open(FLAGS.objects_written_file, 'r') as object_file:
+    object_records = json.load(object_file)
```
I think you need to raise error if object_records is empty, or fewer than the number of workers.

noahl replied:
I changed the logic to a different and much more concise pattern, which should fix this issue.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -518,6 +643,240 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
+    object_records = json.load(object_file)
+
+  num_workers = FLAGS.num_streams
+  objects_by_worker = []
```
please extract this logic out into a function, and write unit-test for it. 

Please decide if you plan to support the case where num-workers > object_records, if so, then current logic would fail. (based on my cursory simulation run in my head for the case of object_records = 3 and num_workers=4), you will run into array out of bounds exception.

noahl replied:
I found a different and much more concise way of writing this function. I can break into a subroutine if you want, but it's only one line now:

`objects_by_worker = [object_records[i::num_workers]
                       for i in xrange(num_workers)]`

I also did some experiments, and that should handle the case where num_workers > len(object_records).

stfeng replied:
Nice simplification. I trust you it's going to work! ;)

Would be nice to put some comments there to show how this is going to work, perhaps with a couple of examples.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -518,6 +643,240 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
+
+  results = RunThreadedWorkers(
+      WriteWorker,
+      (storage_schema,
```
where and when do you pass in the bucket?! (which is expected by the WriteWorker?)

Have you tried to run this code at all?

stfeng replied:
Sorry, I was looking at the WriteObjects function, not the write worker. Never mind....

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -518,6 +643,240 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
+
+  now = time.time()
+  sleep_time = when - now
+  if sleep_time > 0.0:
```
sleep_time being 0 should be supported, if sleep time is 0, just skip sleep instead of logging a warning here.

noahl replied:
Done.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -518,6 +643,240 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
+  sizes = []
+
+  size_iterator = SizeDistributionIterator(size_distribution)
+  object_prefix = ('pkb_write_worker_%s' % worker_num)
```
please add a timestamp to the object_prefix, this is to distinguish between multiple invocations (and it's a convention we use when creating object prefix)

noahl replied:
Done.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -518,6 +643,240 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
+      latencies.append(latency)
+      sizes.append(object_size)
+    except Exception as e:
+      logging.info('Caught exception %s while writing object %s' %
```
also log the worker number here

noahl replied:
Done.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -518,6 +643,240 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
+
+  # operation_records is the data we send back to the controller.
+  operation_records = []
+  for result in results:
```
We need to check if the number of successful results returned here by the workers are above FAILURE_TOLERANCE threshold, just like other test cases do. i.e., total number of reported <start_time, latency, size> triples >= FAILURE_TOLERANCE * TOTAL number of objects requested to be written.

Same for reads.

noahl replied:
Fixed. If the number of reads or writes is too low, it raises LowAvailabilityError, just like the other scenarios.

stfeng commented on perfkitbenchmarker/scripts/object_storage_api_tests.py:
```diff
@@ -518,6 +643,240 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
+      start_time, latency = ReadObject(storage_schema, FLAGS.bucket,
+                                       name, host_to_connect=host_to_connect)
+    except Exception as e:
+      logging.info('Caught exception %s while reading object %s' %
```
log worker number here too.

noahl replied:
Done.

stfeng commented:
I finished another pass, nothing major this time, a couple of bugs though, please take a look. 
(It's getting really close to the finish line now!!)

At the next update, please outline the set of verification and regression tests that you will do after the code is signed-off but before the final merge.

noahl commented:
@stfeng I've updated in response to your comments; please take a look.

For verification and regression testing, I suggest that I run these tests:
  1) Make sure the non-multistream API benchmarks still work and still return similar (within normal variance) results to what they did before the change
  2) Test that the multistream test with 1 stream has similar throughput to the single stream throughput test

I think those are good for regression testing. I'll think more about verification testing.

stfeng commented:
I finished another pass, getting even closer! To summarize the remaining comments:

1. Move the usage of memory view object and the size_to_write variable one level up from WriteObjectFromBuffer(), and share the memory view object.
2. After obtaining raw results from the target VM, we need to aggregate raw results as part of the "analysis" step, and report the final results as Samples (for multiple reasons, see my comments inline for details).

stfeng commented:
@noahl Regarding test plan: the regression test plan you wrote above looks good. As to verification tests for this particular PR, we just need to make sure basic p0 use cases work (e.g., multiple streams + x objects per stream). Since you will have a couple of more PRs for this work (distribution support, post-analysis logic, and multiple VM), more comprehensive verification can happen at later PRs.

noahl commented:
@stfeng I've responded to all of your comments. I left one question for you, about whether to use raw latency or latency / size, but otherwise everything is ready here.

I'm glad you think the verification tests I suggested are enough.

stfeng commented:
@noahl, I've responded to your question above about raw latency aggregation. 

Look forward to review your new commit whenever it's ready. Please do *not* squash, it's easier to review incremental commits. Thanks!

noahl commented:
@stfeng I've updated the PR in response to your review. Please take a look.

noahl commented:
@stfeng I've added a commit in response to your memoryview comments. For right now it just removes the memoryview code. As we discussed in person, I haven't found a way to make it actually zero-copy, even by switching to `set_contents_from_file`.

stfeng commented:
LGTM. 

Please make sure the regression tests and verification tests as discussed above are performed before the merge. Also, I'd run at least one of the test cases on all 3 providers.

noahl commented:
@stfeng I added the TODO you asked for and I've been running tests. Here are the results:

- api_multistream, api_data, and api_namespace scenarios all complete successfully on GCP.
- I ran the api_multistream scenario with 1B objects to compare with the one byte upload scenario. The upload latencies were almost identical. The multistream download latency was about 0.01 seconds slower than the one byte download latency. I think this is noise - they are effectively the same.
- I also checked whether the single-stream upload and download benchmark from the api_data scenario matches the multistream throughput benchmark running with 1 stream and 100MiB objects. The throughputs are generally within 10-20 Mbps of each other at the same percentile. I think that's acceptable, especially given that I'm computing the throughput of the multistream throughput benchmark by inverting the latency numbers, which is very numerically unstable. When I submit the analysis PR I'll try to do better than this. (As further evidence that the number should be the same, the measurements are done by literally the same function in the backend code.)
- api_data scenario completes successfully on AWS and Azure. I didn't check the api_namespace scenario because it takes a very long time to do the test and I didn't touch any code that it uses anyway.

Given these positive results, I am going to go ahead and merge.
